include/picongpu/fields/laserProfiles/laserExpRampWithPrepulse.hpp

```diff
@@ -0,0 +1,198 @@
+ * Be careful - problematic for few cycle pulses. Thought the rest is cloned
+ * from laserWavepacket, the correctionFactor is not included (this made a
+ * correction to the laser phase, which is necessary for very short pulses,
+ * since otherwise the field is not a solution to the the free Maxwell Eq.; so
+ * that in particular a test particle is, after the laser pulse has passed,
+ * not returned to immobility, as it should. Since the analytical solution is
+ * only implemented for the gaussian regime, and we have mostly exponential
+ * regimes here, it was not retained here.
+ */
+
+namespace laserExpRampWithPrepulse
+{
+    constexpr float_X laserTimeShift = laser::initPlaneY * CELL_HEIGHT /
+        SPEED_OF_LIGHT;
+    constexpr float_64 f = SPEED_OF_LIGHT / WAVE_LENGTH;
+    constexpr float_64 w = 2.0 * PI * f;
+    constexpr float_X endUpramp = TIME_PEAKPULSE -
+        0.5 * LASER_NOFOCUS_CONSTANT;
+    constexpr float_X startDownramp = TIME_PEAKPULSE +
+        0.5 * LASER_NOFOCUS_CONSTANT;
+
+    /** takes time t relative to the center of the gaussian and returns value
+     * between 0 and 1, i.e. as multiple of the max value.
+     * use as: amp_t = amp_0 * gauss( t - t_0 )
+     */
+    HDINLINE float_X
+    gauss( float_X const t )
+    {
+        float_X const exponent = t / float_X( PULSE_LENGTH );
+        return math::exp( float_X( -0.25 ) * exponent * exponent );
+    }
+
+    // get value of exponential curve through two points at given t
+    // t/t1/t2 given as float_X, since the envelope doesn't need the accuracy
+    HDINLINE float_X
+    extrapolate_expo(
+        float_X const t1,
+        float_X const a1,
+        float_X const t2,
+        float_X const a2,
+        float_X const t
+    )
+    {
+        const float_X log1 = ( t2 - t ) * math::log( a1 );
+        const float_X log2 = ( t - t1 ) * math::log( a2 );
+        return math::exp( ( log1 + log2 )/( t2 - t1 ) );
+    }
+
+    HINLINE float_X
+    get_envelope(float_X runTime)
+    {
+        float_X env = 0.0;
+        const bool before_preupramp = ( float_X( -0.5 * RAMP_INIT *
+            PULSE_LENGTH ) > runTime );
+        const bool before_start = runTime < 0.;
+        const bool before_peakpulse = runTime < endUpramp;
+        const bool during_first_exp = ( TIME_1 < runTime ) &&
+            ( runTime < TIME_2 );
+        const bool after_peakpulse = startDownramp <= runTime;
+
+        if ( ! before_preupramp && before_start )
+        {
+            env = AMP_1 * gauss( runTime ) +
+                AMP_PREPULSE * gauss( runTime - TIME_PREPULSE );
+        }
+        else if ( before_peakpulse )
+        {
+            const float_X ramp_when_peakpulse = extrapolate_expo(
+                TIME_2,
+                AMP_2,
+                TIME_3,
+                AMP_3,
+                endUpramp
+            ) / AMPLITUDE;
+            // if (ramp_when_peakpulse > 0.5) - I know, dead code :) didn't understand your comment about the other logging
```

Comment:
if this function is only host-code, you can throw an error for now... but this will result in a non-clean shutdown of the MPI & CUDA context...

Reply:
As an alternative, you can instead just log it via:

```C++
log< picLog::PHYSICS >(
    "Attention, the intensities of the ramp are very large!\n"
    "The extrapolation to ..."
);
```

I would actually prefer that instead of throwing an exception here.

Reply:
and Rene told me to try PMACC_ASSERT?

Reply:
that's an alternative but I think the default-on `PHYSICS` log level will do!

Reply:
ot discussion: we are additionally able to express it as compile time check via `PMACC_CASSERT`.